### PR TITLE
[WIP] Generic script crate

### LINKED ISSUE #1799: Split up the script crate for build speed
It's by far the largest crate in terms of binary size (2.5× as big as `main`).  Building it takes almost 40 seconds and 1.8 GB resident on my machine (with debugging enabled, optimization disabled).

The various DOM types and their generated bindings mostly don't depend on each other.  We could split them among a few crates to improve compile time (especially in parallel).

### PR DESCRIPTION
**Pull request by hrvolapeter**
Separation is based on adding types to TypeHolderTrait which contains
associated types with traits that script crate actually needs. This allows
using static methods and Sized types.

Affects servo in two ways:
- generic structs do not compile in script crate but are left for later monomorphization
- dom_structs itself are moved to separate crate

TODO:
- [ ] tests
- [ ] performance
- [ ] polyshing

Fixes #1799

r? @jdm

---

- [ ] `./mach build -d` does not report any errors
- [ ] `./mach test-tidy` does not report any errors
- [ ] These changes fix #1799 (github issue number if applicable).

- [ ] There are tests for these changes OR
- [ ] These changes do not require tests because _____

---
This change is [Reviewable](https://reviewable.io/reviews/servo/servo/21371)

### CODE REVIEW

**Comment by jdm** [comment]
Results from local measurements using non-incremental builds:

master:
`./mach build -d; touch components/script/dom/servoparser/html.rs && ./mach build -d`: 3:49m

PR:
`./mach build -d; touch components/script/fetch.rs && ./mach build -d`: 4:18m
`./mach build -d; touch components/script_servoparser/html.rs && ./mach build -d`: 1:59m

**Comment by jdm** [comment]
Results from local measurements using incremental builds:

master:
`./mach build -d; touch components/script/dom/servoparser/html.rs && ./mach build -d`: 3:19m

PR:
`./mach build -d; touch components/script/fetch.rs && ./mach build -d`: 2:56m
`./mach build -d; touch components/script_servoparser/html.rs && ./mach build -d`: 1:48m

**Comment by jdm** [comment]
@retep007 Does this PR contain the most recent changes you have been making? I might try to keep investigating this week.

**Comment by retep007** [comment]
@jdm It contains all latest changes, however I can rebase it for you. Also I don't really know how to solve the ICE problem, no response from rust-lang so far. Those const functions are used to generate const variables and one of them can't be replaced so easily.

**Comment by jdm** [comment]
Ah, that's good to know. Have you tried making duplicating the function and making a non-const version?

**Comment by retep007** [comment]
Last static fn which I wasn't able to remove is `Guard::new`which I think is beyond manual replacement. But maybe we could try `lazy_static`?

**Comment by jdm** [comment]
I'm not sure I understand - what's the problem with Guard::new, exactly?

**Comment by retep007** [comment]
@jdm I solved the `const fn` issue and rebased I get `window is undefined`. I guess that it origniates from mozjs which is checking for null pointer provided by servo. Maybe something wrong with initialization?

File: components/script_servoparser/domparser.rs [hunk 1]
```diff
@@ -0,0 +1,116 @@
...
+use script::dom::document::{Document, HasBrowsingContext, IsHTMLDocument};
+use script::dom::document::DocumentSource;
+use script::dom::domparser::DOMParserTrait;
+use script::dom::servoparser::ServoParser;
+use script::dom::window::Window;
+use script::typeholder::TypeHolderTrait;
+use script_traits::DocumentActivity;
+
+#[derive(DenyPublicFields, DomObject, JSTraceable, MallocSizeOf)]
+#[base = "script"]
```
Comment (sagudev) [inline components/script_servoparser/domparser.rs:31]: This

**Comment by jdm** [comment]
Closing in favour of #33416. This work was very informative and we learned a lot from it!

**Comment by hrvolapeter** [comment]
Great to see some results of this 😀

**Pull request closed without merging**